activerecord/test/cases/autosave_association_test.rb

```diff
@@ -517,6 +517,24 @@ def test_invalid_adding_before_save
     assert_not_predicate new_firm, :persisted?
   end
 
+  def test_adding_unsavable_association
+    new_firm = Firm.new("name" => "A New Firm, Inc")
+    client = new_firm.clients.new("name" => "Apple")
+    # Stub the save method of the client to simulate aborting the insert
+    class << client
+      def save(*)
+        false
+      end
+    end
+
+    assert_predicate client, :valid?
+    assert_predicate new_firm, :valid?
+    assert_not new_firm.save
+    assert new_firm, :persisted?
```

Comment:
I'm not happy that `new_firm` is `persisted?`, but this is a side-effect of the insert of `new_firm` succeeding and the database providing an ID, even if the transaction is rolled back later. Thoughts?

Reply:
I think this might be prone to obscure bug inside application

> new_firm.persisted? => true
new_firm.new_record? => false
Firm.exists?(new_firm.id) => false

Should we instead reset the `@new_record` ivar when the transaction is rolled back?

Reply:
If everyone agrees that's the solution, I'm all for it and I can do it, I just didn't want to change the existing behaviour

Reply:
What about the ID and other database-provided attributes?

Reply:
More on this: it looks like there is already code that handles rolling back the attributes

https://github.com/rails/rails/blob/8ae38dab5251613a7e7f5af5289e524abca20efd/activerecord/lib/active_record/transactions.rb#L323-L327

However, it looks like `ActiveRecord::Rollback` is swallowed before it reaches that point. 

https://github.com/rails/rails/blob/8ae38dab5251613a7e7f5af5289e524abca20efd/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L247-L257

I'm absolutely not familiar with this code so I'd really appreciate pointers

Reply:
https://github.com/rails/rails/pull/32862 should fix the rollback behaviour so that `new_firm.persisted?` is false.

Incidentally, the `assert` here will never fail, because `new_firm` will always be truthy; it looks like it should be `assert_predicate` instead.

Reply:
> #32862 should fix the rollback behaviour so that new_firm.persisted? is false.

Awesome! I'll rebase and make the changes when it gets merged.

> Incidentally, the assert here will never fail

😂 that takes me back to #20527

Reply:
😄 

I've also bumped into this problem before: https://github.com/cypriss/mutations/pull/110